core/query/mixingQuery.ml

```diff
@@ -0,0 +1,1551 @@
+              | _ -> assert false
+          end
+    | If (c, body, Concat []) ->
+      (* Turn conditionals into where clauses. We might want to do
+         this earlier on.  *)
+      let c = base index c in
+      let (_, fields, tables, c', os) = select_clause index unit_query body in
+      let c = Sql.smart_and c c' in
+      (false, fields, tables, c, os)
+    | Table (_db, table, _keys, (fields, _, _)) ->
+      (* eta expand tables. We might want to do this earlier on.  *)
+      (* In fact this should never be necessary as it is impossible
+         to produce non-eta expanded tables. *)
+      let var = Sql.fresh_table_var () in
+      let fields =
+        Sql.Fields
+          (List.rev
+            (StringMap.fold
+              (fun name _ fields ->
+                (Sql.Project (var, name), name)::fields)
+              fields
+              []))
+      in
+        (false, fields, [Sql.TableRef(table, var)], Sql.Constant (Constant.Bool true), [])
+    | Singleton _ when unit_query ->
+      (* If we're inside an Sql.Empty or a Sql.Length it's safe to ignore
+         any fields here. *)
+      (* We currently detect this earlier, so the unit_query stuff here
+         is redundant. *)
+      (false, Sql.Fields [], [], Sql.Constant (Constant.Bool true), [])
+    | Singleton (Record fields) ->
+      let fields =
+        Sql.Fields
+          (List.rev
+            (StringMap.fold
+              (fun name v fields ->
+                (base index v, name)::fields)
+              fields
+              []))
+      in
+        (false, fields, [], Sql.Constant (Constant.Bool true), [])
+    | _ -> assert false
+and clause : Sql.index -> bool -> Q.t -> Sql.query =
+  fun index unit_query v -> Sql.Select(select_clause index unit_query v)
+and base : Sql.index -> Q.t -> Sql.base = fun index ->
+  let open Q in
+  function
+    | If (c, t, e) ->
+      Sql.Case (base index c, base index t, base index e)
+    | Apply (Primitive "tilde", [s; r]) ->
+      begin
+        match likeify r with
+          | Some r ->
+            Sql.Apply ("LIKE", [base index s; base index r])
+          | None ->
+              begin
+                let r =
+                  (* HACK:
+                     this only works if the regexp doesn't include any variables bound by the query
+                  *)
+                  Sql.Constant (Constant.String (Regex.string_of_regex (Linksregex.Regex.ofLinks (value_of_expression r))))
+                in
+                  Sql.Apply ("RLIKE", [base index s; r])
+              end
+        end
+    | Apply (Primitive "Empty", [v]) ->
+        Sql.Empty (unit_query v)
+    | Apply (Primitive "length", [v]) ->
+        Sql.Length (unit_query v)
+    | Apply (Primitive f, vs) ->
+        Sql.Apply (f, List.map (base index) vs)
+    | Project (Var (x, _tyx), name) ->
+        Sql.Project (x, name)
+    | Constant c -> Sql.Constant c
+    | Primitive "index" ->
```

Comment:
Perhaps leave this out since we are not currently using "index" in mixing queries?

Reply:
Factorised into `QueryLang` (and we still need to consider "index" there...)